### PR TITLE
[Model] Add HiDream-O1-Image support (text-to-image, editing, personalization, layout-bbox)

### PR DESCRIPTION
**Pull request by SOKOUDJOU-LEOPOLD**
## Purpose

This PR adds support for **HiDream-O1-Image** (`HiDream-ai/HiDream-O1-Image` and `HiDream-ai/HiDream-O1-Image-Dev`), a ~8B-parameter image generation model from HiDream.ai. It supports text-to-image, instruction-based image editing, multi-reference subject personalization, and layout-bbox spatial conditioning — all in a single model with no VAE and no separate text encoder.

### Why

HiDream-O1-Image has an architecture that nothing in vLLM-Omni currently handles:

- **No VAE.** The model diffuses directly on raw 32×32 pixel patches. Output is assembled from final hidden states with no latent encoder/decoder.
- **No separate text encoder.** A single Qwen3-VL backbone handles everything — causal attention over text for conditioning, bidirectional attention over image tokens for the diffusion pass. This is called a Pixel-level Unified Transformer (UiT).
- **Timestep conditioning via token substitution, not AdaLN.** A special `<|tms_token|>` placeholder has its embedding swapped for the sinusoidal timestep embedding at every step.
- **Mixed causal/bidirectional attention in one forward call.** Text tokens are causal; image tokens attend bidirectionally over the full sequence. Expressed through `AttentionMetadata` + `full_attn_spans`.
- **3D MRoPE with `fix_point` anchoring.** Custom `get_rope_index_fix_point` pins reference image positions at a fixed offset to avoid RoPE extrapolation.

### What

**New files — `vllm_omni/diffusion/models/hidream_o1_image/`** (4 files):

- `qwen3_vl_uit_transformer.py` — `HiDreamO1UiTModel` backbone. Lazy-imports HF Qwen3-VL (gated on `transformers >= 4.57.1`). Overrides attention/MLP projections with TP-capable parallel linears via `_build_attention_subclasses(quant_config)` closure. Carries `_cache_dit_adapter_config`, `_sp_plan`, `_hsdp_shard_conditions`, `_layerwise_offload_blocks_attrs`.
- `pipeline_hidream_o1_image.py` — `HiDreamO1ImagePipeline` denoise loop. Handles T2I, editing, personalization, layout-bbox. Implements `SupportsComponentDiscovery` (empty `_vae_modules`, nested `_encoder_modules = ["transformer.visual"]`), `DiffusionPipelineProfilerMixin`, `CFGParallelMixin`.
- `utils_hidream_o1.py` — pure utilities: `build_packed_attention_metadata` (builds both `full_attn_spans` and dense `attn_mask`), `get_rope_index_fix_point`, `patchify`/`depatchify`, layout-bbox rendering.
- `__init__.py` — module exports.

**Parallelism — all wired in the model files above:**

| Strategy | Implementation |
|----------|---------------|
| Tensor Parallel | `QKVParallelLinear` (fused Q/K/V), `MergedColumnParallelLinear` (gate+up), `RowParallelLinear` (o_proj, down_proj). `stacked_params_mapping` in `load_weights()` remaps split checkpoint weights into fused params. |
| Sequence Parallel (Ulysses + Ring) | `_sp_plan` splits `inputs_embeds`/`position_ids` before decoder stack, gathers `final_layer2` after. `attn_metadata` excluded — Ulysses AllToAll restores full sequence before attention. T2I only; editing path documented as known SP limitation. |
| CFG-Parallel | Zero code — inherited from `CFGParallelMixin`. |
| HSDP | `_hsdp_shard_conditions = [_is_transformer_block]` — each `Qwen3VLTextDecoderLayer` is a shard unit. |
| Cache-DiT | `_cache_dit_adapter_config` with `ForwardPattern.Pattern_0`, `has_separate_cfg=True`, `check_forward_pattern=False`. Added `layers` property alias on `HiDreamO1UiTModel` pointing to `language_model.layers` for cache-dit auto-detection. |
| CPU Offload | `SupportsComponentDiscovery`. `_vae_modules = []` (no VAE). `_encoder_modules = ["transformer.visual"]` — first nested encoder path in the codebase, resolved via `operator.attrgetter`. |
| Layerwise Offload | `_layerwise_offload_blocks_attrs = ["language_model.layers"]` |

**Integration:**
- `vllm_omni/diffusion/registry.py` — registered with `get_hidream_o1_image_post_process_func` (`[-1, 1]` → uint8, no VAE decode).
- `docs/models/supported_models.md`, `docs/user_guide/diffusion_features.md`, `docs/user_guide/diffusion/cpu_offload_diffusion.md` — rows added.
- `.buildkite/cuda/test-ready.yml`, `test-merge.yml`, `test-nightly.yml` — CI entries added.
- `examples/offline_inference/hidream_o1_image/`, `examples/online_serving/hidream_o1_image/` — runnable examples with READMEs.
- `recipes/HiDream/HiDream-O1-Image.md` — full serving recipe with two-bucket performance methodology.

## Test Plan

**Unit tests (47 tests, CPU-only, `pytest -m core_model`):**

| File | Coverage |
|------|----------|
| `tests/diffusion/models/hidream_o1_image/test_packed_sequence.py` | `build_packed_attention_metadata` for T2I, editing, multi-ref, layout-bbox. Verifies `full_attn_spans` coordinates, `attn_mask`, token types, causal vs bidirectional pattern. |
| `tests/diffusion/models/hidream_o1_image/test_embed_cache.py` | Vision-tower called exactly once per request across all denoising steps. |
| `tests/diffusion/models/hidream_o1_image/test_layout_bbox_utils.py` | Bbox parsing, layout image rendering, coordinate clamping, batch handling. |

**E2E offline** (`tests/e2e/offline_inference/test_hidream_o1_image.py`): full pipeline forward pass, output shape and pixel range verified.

**E2E online** (`tests/e2e/online_serving/test_hidream_o1_image.py`): T2I (Dev + Full variants), image editing, multi-reference personalization, layout-bbox.

**Nightly expansion** (`tests/e2e/online_serving/test_hidream_o1_image_expansion.py`): 8 parametrized server configs:

| id | flags | GPUs |
|----|-------|------|
| `default` | — | 1× H100 |
| `cache_dit` | `--cache-backend cache_dit` | 1× H100 |
| `tp2` | `--tensor-parallel-size 2` | 2× H100 |
| `ulysses2` | `--ulysses-degree 2` | 2× H100 |
| `cfg_parallel2` | `--cfg-parallel-size 2` | 2× H100 |
| `hsdp` | `--use-hsdp` | 1× H100 |
| `cpu_offload` | `--enable-cpu-offload` | 1× H100 |
| `layerwise_offload` | `--enable-layerwise-offload` | 1× H100 |

**vLLM Version:** 0.23.0

**vLLM-Omni Commit:** 1ff2f22ce21c7e6c3a40919b321f51f48509143e

## Test Result

Unit tests (47): all passed on CPU.

image

E2E and expansion tests require H100 hardware, tracked through CI (test-merge and test-nightly pipelines).

Pixel-level output comparison against the HiDream reference repo is explicitly deferred and tracked in #5304 .
H100 benchmark numbers are explicitly deferred and tracked in #5305 .

### CODE REVIEW

File: vllm_omni/diffusion/models/hidream_o1_image/utils_hidream_o1.py [hunk 1]
```diff
@@ -0,0 +1,582 @@
...
+
+    causal = torch.ones(seq_len, seq_len, dtype=torch.bool, device=device).tril(diagonal=0)
+    dense_mask = causal.unsqueeze(0).repeat(batch_size, 1, 1)
+
+    full_attn_spans: list[list[tuple[int, int]]] = [[] for _ in range(batch_size)]
+    is_gen = token_types > 0
+    for b in range(batch_size):
+        dense_mask[b, is_gen[b]] = True  # gen rows attend to everything (incl. future gen positions)
+        row = is_gen[b]
+        idx = torch.nonzero(row, as_tuple=False).squeeze(-1)
```
[the remark this answers is not in the document]
Comment (SOKOUDJOU-LEOPOLD) [reply vllm_omni/diffusion/models/hidream_o1_image/utils_hidream_o1.py:269]: Invalid. 

It would add a redundant operation (dense_mask[b, ~gen] &= (~gen).unsqueeze(0)) that zeroes bits already guaranteed to be zero by construction. The docstring phrase "causal-only-over-text" is accurate as is, text tokens already only see text columns because the layout puts all gen tokens at the tail of every sequence.

File: vllm_omni/diffusion/models/hidream_o1_image/pipeline_hidream_o1_image.py [hunk 2]
```diff
@@ -0,0 +1,686 @@
+# SPDX-License-Identifier: Apache-2.0
+# SPDX-FileCopyrightText: Copyright contributors to the vLLM project
+"""HiDream-O1-Image pipeline — Phase 3 (text-to-image + image editing +
+multi-reference personalization + layout-bbox conditioning, single GPU,
+no Cache-DiT / TP / SP yet).
+
```
[the remark this answers is not in the document]
Comment (SOKOUDJOU-LEOPOLD) [reply vllm_omni/diffusion/models/hidream_o1_image/pipeline_hidream_o1_image.py]: Valid.
The docstring was written at the end of Phase 3 and still says "single GPU, no Cache-DiT / TP / SP yet", but Phases 5–8 added all of those. It's now stale and misleading.

File: vllm_omni/diffusion/models/hidream_o1_image/pipeline_hidream_o1_image.py [hunk 3]
```diff
@@ -0,0 +1,686 @@
...
+            [[1, h_patches, w_patches]], dtype=torch.int64, device=self.device
+        )
+        ref_pixel_grid_thw = torch.tensor(
+            [[1, pil.size[1] // PATCH_SIZE, pil.size[0] // PATCH_SIZE]
+             for pil, n in zip(ref_pils, ref_image_lens)
+             for _ in [resize_pilimage(pil, ref_max)]],
+            dtype=torch.int64,
+            device=self.device,
+        )
+        # Build per-ref pixel grid from actual resized dimensions.
```
[the remark this answers is not in the document]
Comment (SOKOUDJOU-LEOPOLD) [reply vllm_omni/diffusion/models/hidream_o1_image/pipeline_hidream_o1_image.py]: Valid.

- Lines 388–394 build ref_pixel_grid_thw but with a subtle bug: it calls resize_pilimage(pil, ref_max) inside the comprehension expression (for _ in [resize_pilimage(...)]) but then reads pil.size (the original unresized dimensions), not the resized ones. This produces wrong grid values AND does a wasted resize call.
- Lines 396–401 immediately overwrite ref_pixel_grid_thw correctly: it resizes, captures pil_r.size, and uses rw, rh.

File: examples/offline_inference/hidream_o1_image/text_to_image_hidream_o1.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    parser = argparse.ArgumentParser(description="HiDream-O1-Image text-to-image generation.")
+    parser.add_argument("--model", default="HiDream-ai/HiDream-O1-Image", help="Model name or local path.")
+    parser.add_argument("--prompt", default="A cute cat sitting on a windowsill at sunset.", help="Text prompt.")
+    parser.add_argument("--height", type=int, default=1024, help="Output image height (snapped to a 32px grid).")
+    parser.add_argument("--width", type=int, default=1024, help="Output image width (snapped to a 32px grid).")
+    parser.add_argument(
+        "--model-type",
+        choices=["full", "dev"],
+        default="full",
+        help="full: 50 steps, guidance 5.0, shift 3.0. dev: 28 steps, guidance 0.0 (no CFG), shift 1.0.",
```
[the remark this answers is not in the document]
Comment (SOKOUDJOU-LEOPOLD) [reply examples/offline_inference/hidream_o1_image/text_to_image_hidream_o1.py]: valid

File: vllm_omni/diffusion/models/hidream_o1_image/pipeline_hidream_o1_image.py [hunk 5]
```diff
@@ -0,0 +1,686 @@
...
+        noise_scale = sp.extra_args.get("noise_scale", DEFAULT_NOISE_SCALE)
+        seed = sp.seed if sp.seed is not None else 42
+
+        # Layout-bbox conditioning (Phase 3): if layout_bboxes are provided,
+        # augment ref_pils with bordered versions + a layout image.  The
+        # augmented list is then handled identically to regular multi-ref input.
+        raw_layout = mm_data.get("layout_bboxes")
+        if raw_layout is not None and ref_pils is not None:
+            layout_data = load_layout_bboxes(raw_layout) if isinstance(raw_layout, str) else raw_layout
+            ref_pils = create_layout_reference_images(ref_pils, layout_data, width, height)
```
[the remark this answers is not in the document]
Comment (SOKOUDJOU-LEOPOLD) [reply vllm_omni/diffusion/models/hidream_o1_image/pipeline_hidream_o1_image.py:662]: Valid.
load_layout_bboxes calls os.path.exists() on any user-supplied string in online serving a remote client controls multi_modal_data["layout_bboxes"], so they can probe server-local file existence and read arbitrary JSON files from disk.
To fix:
- use json.loads() directly, only accept inline JSON, never read files from user data.
- preload the file with load_layout_bboxes() before putting data in multi_modal_data, since that's operator-controlled code anyway.

**Comment by SOKOUDJOU-LEOPOLD** [comment]
Hi @hsliuustc0106 , Please can you review my PR?

File: examples/online_serving/hidream_o1_image/run_server.sh [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+# Set MODEL to HiDream-ai/HiDream-O1-Image for the full 50-step variant.
+
+MODEL="${MODEL:-HiDream-ai/HiDream-O1-Image-Dev}"
+PORT="${PORT:-8095}"
+
+echo "Starting HiDream-O1-Image server..."
+echo "Model: $MODEL"
+echo "Port:  $PORT"
+
+vllm serve "$MODEL" --omni \
```
Comment (hsliuustc0106) [inline examples/online_serving/hidream_o1_image/run_server.sh:13]: [P1] Add automatic model-class resolution before relying on this command. The official checkpoint config declares `Qwen3VLForConditionalGeneration`, but the resolver returns that unregistered architecture while the registry contains `HiDreamO1ImagePipeline`, so `vllm serve "$MODEL" --omni` will fail unless the caller supplies `--model-class-name`. Please add the HiDream signature mapping to both resolver/enrichment paths; carrying #5194's diffusion detector alone is not sufficient because detection does not select the pipeline class.

**Review by hsliuustc0106 — COMMENTED** [review]
Joint review with #5194. These are comment-only findings; no approval or change-request verdict.

File: vllm_omni/diffusion/models/hidream_o1_image/qwen3_vl_uit_transformer.py [hunk 7]
```diff
@@ -0,0 +1,694 @@
...
+    # along the sequence dimension (dim 1) and position_ids along its sequence
+    # dimension (dim 2, since shape is (3, B, S)).  attn_metadata is NOT in the
+    # plan and passes through unchanged: Ulysses AllToAll reconstructs the full
+    # sequence on each rank before local attention, so full_attn_spans and
+    # attn_mask global coordinates remain correct.
+    #
+    # Limitation: the reference-image path (pixel_values != None) uses
+    # visual_pos_masks for deepstack indexing into the sequence.  Those
+    # positions may span multiple SP ranks, making gather/scatter logic
+    # non-trivial.  For now SP is supported for text-to-image only; using
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/models/hidream_o1_image/qwen3_vl_uit_transformer.py:500]: [P1] Guard this combination before advertising sequence parallelism for the full pipeline. With reference images, `visual_pos_masks` remains full-length while `_sp_plan` splits only `inputs_embeds` and `position_ids`; `_deepstack_process()` then indexes local hidden states with the full mask. Edit/personalization/layout with `--ulysses-degree > 1` can therefore shape-error or produce incorrect deepstack embeddings. Please either implement mask-aware split/gather or reject this mode explicitly and add a reference-image SP test.

File: examples/online_serving/hidream_o1_image/run_curl_hidream_o1.sh [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+HOST="${HOST:-localhost}"
+PORT="${PORT:-8095}"
+OUTPUT="${OUTPUT:-hidream_o1_output.png}"
+
+curl -s "http://${HOST}:${PORT}/v1/images/generations" \
+    -H "Content-Type: application/json" \
+    -d '{
+        "prompt": "A golden retriever running through a field of sunflowers at sunset",
+        "size": "1024x1024",
+        "num_inference_steps": 28,
```
Comment (hsliuustc0106) [inline examples/online_serving/hidream_o1_image/run_curl_hidream_o1.sh:14]: [P1] Apply the documented Dev sampling defaults to this online path. The server defaults to the Dev variant and this request sends 28 steps, but omits guidance and shift; the pipeline consequently substitutes guidance `5.0` and shift `3.0`, enabling CFG and using the full-model shift instead of the Dev preset (`1.0`/`1.0`). Please make defaults model-variant-aware or include and enforce the Dev parameters in the server example.

File: vllm_omni/diffusion/models/hidream_o1_image/pipeline_hidream_o1_image.py [hunk 9]
```diff
@@ -0,0 +1,679 @@
...
+                if do_true_cfg and negative_sample is not None:
+                    neg_vinputs = torch.cat([z, ref_patches], dim=1) if ref_patches is not None else None
+                    negative_kwargs: dict[str, Any] | None = {
+                        **negative_sample,
+                        "z": z,
+                        "t_pixeldit": t_pixeldit,
+                        "sigma": sigma,
+                        "vinputs": neg_vinputs,
+                        # CFG uncond branch reuses the same cached embeds (same refs)
+                        "precomputed_image_embeds": cached_img_emb,
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/models/hidream_o1_image/pipeline_hidream_o1_image.py:595]: [P2] Refresh the vision-embedding cache before constructing the negative CFG branch. `cached_img_emb` and `cached_ds_emb` are captured before the positive branch populates `embed_storage`; the negative branch still receives those stale `None` values and `_embed_storage=None`, so reference-image CFG recomputes the vision tower on the first denoise step. Please refresh the cached values after the positive call and add an actual pipeline call-count test.
